File: 08 linkedlists/linkedlist.cs
class LinkedListItem
{
  public int Value { get; set; }
  public LinkedListItem Next { get; set; }
  public LinkedListItem Prev { get; set; }
  public LinkedListItem(int value, LinkedListItem Next)
  {
    this.Value = value;
    this.Next = null;
    this.Prev = null;
  }
}




class LinkedList
{
  public LinkedListItem Head { get; protected set; }
  public LinkedListItem Tail { get; protected set; }
  public int Length { get; protected set; }


  public LinkedList(LinkedListItem head)
  {
    this.SetHead(head);
  }

  public void SetHead(LinkedListItem head)
  {
    this.Head = this.Tail = null;
    this.Length = 0;

    if (head != null)
    {
      this.Head = head;
      this.Length = 1;

      var item = head;
      while (item.Next != null)
      {
        item = item.Next;
        this.Length++;
      }

      this.Tail = item;
    }
  }

  public void Append(LinkedListItem item)
  {
    if (this.Tail != null)
    {
      this.Tail.Next = item;
      item.Prev = this.Tail;
      this.Tail = item;
      this.Length++;
    }
    else
    {
      this.SetHead(item);
    }
  }

  public void Prepend(LinkedListItem item)
  {
    if (this.Head != null)
    {
      item.Next = this.Head;
      this.Head.Prev = item;
      this.Head = item;
      this.Length++;
    }
    else
    {
      this.SetHead(item);
    }
  }

  public LinkedListItem TraverseTo(int index)
  {
    var current = this.Head;
    var i = 0;
    while (i < index)
    {
      current = current.Next;
      i++;
    }

    return current;
  }

  public void Insert(int index, LinkedListItem item)
  {
    if (index == 0)
    {
      this.Prepend(item);
      return;
    }

    if (index >= this.Length)
    {
      this.Append(item);
      return;
    }


    var current = this.TraverseTo(index - 1);

    item.Next = current.Next;
    item.Prev = current;
    current.Next = item;
    item.Next.Prev = item;
    this.Length++;
  }

  public void Delete(int index)
  {
    if (0 == index)
    { // head
      this.Head = this.Head.Next;
      this.Length--;

      if (this.Length == 1)
        this.Tail = this.Head;
      else if (this.Length == 0)
        this.Tail = null;
      else
        this.Head.Prev = null;

      return;
    }

    var current = this.TraverseTo(index);

    current.Prev.Next = current.Next;
    if (current.Next != null)
    {
      current.Next.Prev = current.Prev;
    }

    if (this.Length - 1 == index)
    {
      this.Tail = current.Prev;
    }

    this.Length--;
  }

  public int[] values()
  {
    var item = this.Head;

    var values = new int[this.Length];
    var idx = 0;
    while (item != null)
    {
      values[idx++] = item.Value;
      item = item.Next;
    }

    return values;
  }

  public void reverse()
  {
    var current = this.Tail;
    while (current != null)
    {
      var tmpPrev = current.Prev;
      var tmpNext = current.Next;

      current.Next = current.Prev;
      current.Prev = tmpNext;

      current = tmpPrev;
    }

    var tmpTail = this.Tail;
    this.Tail = this.Head;
    this.Head = tmpTail;
  }

  public void print()
  {
    var current = this.Head;
    while (current != null)
    {
      Console.WriteLine("{0} --> ( {1} ) --> {2}",
        current.Prev != null ? current.Prev.Value : new int?(),
        current.Value,
        current.Next != null ? current.Next.Value : new int?());

      current = current.Next;
    }
  }
}

File: 08 linkedlists/linkedlist.py
class LinkedListItem (object):
  def __init__(self, value, next=None):
    self.value = value
    self.next = None
    self.prev = None


class LinkedList (object):
  def __init__(self, head):
    self.head = None
    self.tail = None
    self.length = 0
    self._setHead(head)

  def _setHead(self, head):
    self.head = self.tail = None
    self.length = 0

    if (not head is None):
      self.head = head
      self.length = 1

      item = head
      while (item.next):
        item = item.next
        self.length += 1

      self.tail = item

  def append(self, item):
    if (self.tail):
      self.tail.next = item
      item.prev = self.tail
      self.tail = item
      self.length += 1
    else:
      self._setHead(item)

  def prepend(self, item):
    if (self.head):
      item.next = self.head
      self.head.prev = item
      self.head = item
      self.length += 1
    else:
      self._setHead(item)

  def traverseTo(self, index):
    current = self.head
    i = 0
    while (i < index):
      current = current.next
      i += 1

    return current

  def insert(self, index, item):
    if (index == 0):
      self.prepend(item)
      return

    if (index >= self.length):
      self.append(item)
      return

    current = self.traverseTo(index - 1)

    item.next = current.next
    item.prev = current
    current.next = item
    item.next.prev = item
    self.length += 1

  def delete(self, index):
    if (0 == index):
      self.head = self.head.next
      self.length -= 1

      if (self.length == 1):
        self.tail = self.head
      elif (self.length == 0):
        self.tail = None
      else:
        self.head.prev = None

      return

    current = self.traverseTo(index)

    current.prev.next = current.next
    if (current.next):
      current.next.prev = current.prev

    if (self.length - 1 == index):
      self.tail = current.prev

    self.length -= 1

  def values(self):
    item = self.head

    values = []
    while (item):
      values.append(item.value)
      item = item.next

    return values

  def reverse(self):
    current = self.tail
    while (current):
      tmpPrev = current.prev
      tmpNext = current.next

      current.next = current.prev
      current.prev = tmpNext

      current = tmpPrev

    tmpTail = self.tail
    self.tail = self.head
    self.head = tmpTail

  def print(self):
    current = self.head
    while (current):
      print("%d --> ( %d ) --> %d",
            current.prev.value if not current.prev is None else "",
            current.value,
            current.next.value if not current.next is None else "")

      current = current.next


lst = LinkedList(None)
lst.append(LinkedListItem(3))
lst.append(LinkedListItem(4))
lst.append(LinkedListItem(5))
lst.prepend(LinkedListItem(1))
lst.insert(1,  LinkedListItem(2))
lst.reverse()
print(lst.values())
lst.delete(0)
print(lst.values())
